src/System.Private.CoreLib/shared/System/Collections/ObjectModel/Collection.cs

```diff
@@ -162,6 +174,47 @@ protected virtual void SetItem(int index, T item)
             items[index] = item;
         }
 
+        protected virtual void InsertItemsRange(int index, IEnumerable<T> collection)
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
+
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if (index < 0 || index > items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            int i = 0;
+            foreach (var item in collection)
+            {
+                Insert(index + i, item);
+                i++;
+            }
+        }
+
+        protected virtual void RemoveItemsRange(int index, int count)
+        {
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if (index < 0 || index > items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            for (int i = index; i < (index + count); i++)
+                RemoveAt(index);
```

Comment:
nit: add braces

Reply:
Should this be `RemoveAt(i)`?

Reply:
Yes, it should

Reply:
After updating the code here to use `RemoveAt(i)` many of the RemoveRange unit tests started failing. I took a closer look at what is happening and I believe `RemoveAt(index)` is correct. 

Consider the following array `[ 1, 2, 3, 4 ]`

Let's say the code invokes `RemoveRange(0, 2);` that means in the scope of the for loop is:
* `index = 0`
* `length = 2`

*Step 1:*
`i = 0`
`RemoveAt(i)`
Result: `[ 2, 3, 4 ]`

*Step 2:*
`i = 1`
`RemoveAt(i)`
Result: `[ 2, 4 ]`

*Final Result:*
Expected: `[ 3, 4 ]`
Actual: `[ 2, 4 ]`

## Conclusion
Since the array is being removed while we iterate through the for loop it is best to use `RemoveAt(index)` which will stay constant as we iterate through the `length` of items we want to remove.

Reply:
> Since the array is being removed while we iterate through the for loop it is best to use `RemoveAt(index)` which will stay constant as we iterate through the `length` of items we want to remove.

Intuitively (i.e. without diving into the implementation), that's what I assumed. It looked odd given the loop condition, so I brought up the question. Thanks for the investigation and clarification. :)

Reply:
Given that conclusion, I think we can get away with a simpler loop condition:
```C#
for (int i = 0; i < count; i++)
{
   RemoveAt(index);
}
```

Reply:
@ahsonkhan I might have missed a comment, did we say we wanted to throw an exception if `index + count > items.Count`? Currently it validates that `count >= 0` but it doesn't do any other validation on `count`. No problem to add it in, just want to clarify before I make the change

Would that be `ArgumentOutOfRangeException`

Reply:
It's the comment below: https://github.com/dotnet/coreclr/pull/23018#discussion_r262361518

> I think we should follow what `List<T>.RemoveRange` does, which is to throw an `ArgumentException`:

Reply:
Thanks for the clarification